rest-service/manager_rest/deployment_update/step_extractor.py

```diff
@@ -0,0 +1,306 @@
+        # role as a part of the EntityID, and to add another parameter,
+        # `entity_type`, that will eliminate the need for `_to_singular_from`.
+
+        self.entity_id.append(entities_id)
+        for entity_name in new_entities:
+            self.entity_id.append(entity_name)
+            if entity_name in old_entities:
+                if old_entities[entity_name] != new_entities[entity_name]:
+                    self._create_deployment_update_step(
+                        _to_singular_form(entities_id), modify=True)
+            else:
+                self._create_deployment_update_step(
+                    _to_singular_form(entities_id))
+
+            del self.entity_id[-1]
+        del self.entity_id[-1]
+
+    def _extract_steps(self, new, old):
+
+        for entities_name, new_entities in new.iteritems():
+            old_entities = old.get(entities_name, {})
+
+            if entities_name == NODES:
+                self._extract_steps_from_nodes(
+                    new_entities, old_entities)
+
+            elif entities_name == OUTPUTS:
+                self._extract_steps_from_entities(
+                    OUTPUTS, new_entities, old_entities)
+
+            elif entities_name == WORKFLOWS:
+                self._extract_steps_from_entities(
+                    WORKFLOWS, new_entities, old_entities)
+
+    def _create_deployment_update_step(self, entity_type, modify=False):
+
+        entity_id_string = str(self.entity_id)
+
+        if modify and not self._flip_diff_perspective:
+            self.deployment_update_manager.create_deployment_update_step(
+                self.deployment_update_id,
+                'modify',
+                entity_type,
+                entity_id_string)
+
+        # Notice that you can identify a modification whether you compare the
+        # old plan to the new plan, or whether you compare new plan to the old
+        # plan, but you only need to create one step of modification.
+        # We chose, arbitrarily, to create the modification step when we
+        # compare the new to the old (_flip_diff_perspective == False).
+        elif modify and self._flip_diff_perspective:
+            return
+
+        else:
+            action = 'add' if not self._flip_diff_perspective else 'remove'
+
+            self.deployment_update_manager.create_deployment_update_step(
+                self.deployment_update_id,
+                action,
+                entity_type,
+                entity_id_string)
+
+
+def extract_steps(deployment_update):
+    steps_extractor = \
+        DeploymentUpdateStepsExtractor(deployment_update)
+
+    # update the steps extractor with the old and the new deployment plans
+    steps_extractor.old_deployment_plan = \
+        DeploymentPlan.from_storage(steps_extractor.deployment_id)
+    steps_extractor.new_deployment_plan = \
+        DeploymentPlan.from_deployment_update(deployment_update)
+
+    steps_extractor.extract_steps()
```

Comment:
i dont like the side affects on this; i think i'd either extract all steps then do the actual changes here, or even at the caller's side